File: src/main/java/com/frank/book/BookRegister.java
package com.frank.book;

import java.util.HashMap;
import java.util.Map;

public class BookRegister {

    final private Map<Object, Book> register;

    final private Map<Book, Availability> availability;

    public BookRegister() {
        register = new HashMap<>();
        availability = new HashMap<>();
    }

    public void add(Book item) {
        availability.put(item, Availability.ONE);
        register(item);
    }

    public Book get(Object key) {
        return register.get(key);
    }

    public void setBookAsOutOfOrder(Book item) {
        this.availability.put(item, Availability.ZERO);
    }

    public void setBookAvailable(Book item) {
        if (this.availability.containsKey(item)) {
            this.availability.get(item).increase();
        } else {
            this.availability.put(item, Availability.ONE);
        }
    }

    public boolean isBookAvailable(Book item) {
        return this.availability.get(item).getAvailability() > 0;
    }

    public void clean() {
        register.clear();
    }

    private void register(Book item) {
        if (!register.containsKey(item.getId())) {
            register.put(item.getId(), item);
        }
    }

}

File: src/main/kotlin/com/frank/entity/book/BookRegister.kt
package com.frank.entity.book

import java.math.BigDecimal
import java.util.HashMap

class BookRegister {

    private val register: MutableMap<Any, Book>

    private val availability: MutableMap<Book, BookAvailability>

    init {
        register = HashMap()
        availability = HashMap()
    }

    fun add(item: Book) {
        availability[item] = BookAvailability.ONE
        register(item)
    }

    operator fun get(key: Any): Book? {
        return register[key]
    }

    fun setBookAsOutOfOrder(item: Book) {
        this.availability[item] = BookAvailability.ZERO
    }

    fun setBookAvailable(item: Book) {
        if (this.availability.containsKey(item)) {
            this.availability[item]?.increase()
        } else {
            this.availability[item] = BookAvailability.ONE
        }
    }

    fun isBookAvailable(item: Book): Boolean {
//        TODO 2 can't use the operator while accessing nullable fields...
//        Infix call corresponds to a dot-qualified call 'this.availability[item]?.availability.compareTo(0)'
//          which is not allowed on a nullable receiver 'this.availability[item]?.availability'. Use '?.'-qualified call instead
//        return this.availability[item]?.availability > 0
        return this.availability[item]?.availability?.compareTo(0) != 0
    }

    fun clean() {
        register.clear()
    }

    private fun register(item: Book) {
        if (!register.containsKey(item.id)) {
            register[item.id] = item
        }
    }

}
